arroyo-api/src/udfs.rs

```diff
@@ -0,0 +1,248 @@
+use crate::queries::api_queries;
+use crate::queries::api_queries::{
+    CreateUdfParams, DbUdf, DeleteUdfParams, GetUdfByNameParams, GetUdfParams,
+};
+use crate::rest::AppState;
+use crate::rest_utils::{
+    authenticate, client, log_and_map, not_found, ApiError, BearerAuth, ErrorResp,
+};
+use crate::to_micros;
+use arroyo_rpc::api_types::udfs::{
+    GlobalUdf, UdfLanguage, UdfPost, UdfValidationResult, ValidateUdfPost,
+};
+use arroyo_rpc::api_types::GlobalUdfCollection;
+use arroyo_rpc::grpc::controller_grpc_client::ControllerGrpcClient;
+use arroyo_rpc::grpc::{CheckUdfsReq, ValidationResult};
+use arroyo_rpc::public_ids::{generate_id, IdTypes};
+use axum::extract::{Path, State};
+use axum::Json;
+use axum_extra::extract::WithRejection;
+use cornucopia_async::Params;
+use tracing::warn;
+
+impl Into<GlobalUdf> for DbUdf {
+    fn into(self) -> GlobalUdf {
+        GlobalUdf {
+            id: self.pub_id,
+            name: self.name,
+            created_at: to_micros(self.created_at),
+            language: self.language.into(),
+            definition: self.definition,
+            updated_at: to_micros(self.updated_at),
+            description: self.description,
+        }
+    }
+}
+
+/// Create a global UDF
+#[utoipa::path(
+    post,
+    path = "/v1/udfs",
+    tag = "udfs",
+    request_body = UdfPost,
+    responses(
+        (status = 200, description = "Created UDF", body = Udf),
+    ),
+)]
+pub async fn create_udf(
+    State(state): State<AppState>,
+    bearer_auth: BearerAuth,
+    WithRejection(Json(req), _): WithRejection<Json<UdfPost>, ApiError>,
+) -> Result<Json<GlobalUdf>, ErrorResp> {
+    let client = client(&state.pool).await.unwrap();
+    let auth_data = authenticate(&state.pool, bearer_auth).await.unwrap();
+
+    let pub_id = generate_id(IdTypes::Udf);
```

Comment:
Should we be validating the UDF before we create it? It looks like right now we rely on the console / api consumer to do the right thing. It would also be good to do this within a transaction with the creation so that we can avoid race co conditions (like two requests to create the same UDF both succeeding).

Reply:
I've made this transactional and added a unique constraint on the name in the database (this required splitting out a `prefix` column, so you'll need to reapply the migration). In the UI it presents like this:

<img width="399" alt="image" src="https://github.com/ArroyoSystems/arroyo/assets/8881183/533a5c4d-015a-41f9-8003-7a1436cbfe6d">

If we want to do the full validation as part of this transaction then we need to decide what to do about _failure to validate_ errors, like when the compiler service isn't running.